zkevm-circuits/src/witness/mpt.rs

```diff
@@ -0,0 +1,190 @@
+                let first = rows.next().unwrap();
+                let last = rows.last().unwrap_or(first);
+                (
+                    key,
+                    MptUpdate {
+                        key,
+                        old_root: Word::from(i as u64),
+                        new_root: Word::from(i as u64 + 1),
+                        old_value: value_prev(first),
+                        new_value: value(last),
+                    },
+                )
+            })
+            .collect();
+        MptUpdates(map)
+    }
+
+    pub(crate) fn table_assignments<F: Field>(&self, randomness: F) -> Vec<MptUpdateRow<F>> {
+        self.0
+            .values()
+            .map(|update| {
+                let (new_root, old_root) = update.root_assignments(randomness);
+                let (new_value, old_value) = update.value_assignments(randomness);
+                MptUpdateRow([
+                    update.key.address(),
+                    update.key.storage_key(randomness),
+                    update.key.field_tag(),
+                    new_root,
+                    old_root,
+                    new_value,
+                    old_value,
+                ])
+            })
+            .collect()
+    }
+}
+
+impl MptUpdate {
+    pub(crate) fn value_assignments<F: Field>(&self, word_randomness: F) -> (F, F) {
+        let assign = |x: Word| {
+            let mut result =
+                RandomLinearCombination::random_linear_combine(x.to_le_bytes(), word_randomness);
+            if let Key::Account { field_tag, .. } = self.key {
+                if matches!(field_tag, AccountFieldTag::Nonce) {
+                    result = x.to_scalar().unwrap()
+                }
+            }
+            result
+        };
+
+        (assign(self.new_value), assign(self.old_value))
+    }
+
+    pub(crate) fn root_assignments<F: Field>(&self, word_randomness: F) -> (F, F) {
+        (
+            RandomLinearCombination::random_linear_combine(
+                self.new_root.to_le_bytes(),
+                word_randomness,
+            ),
+            RandomLinearCombination::random_linear_combine(
+                self.old_root.to_le_bytes(),
+                word_randomness,
+            ),
+        )
+    }
+}
+
+#[derive(Eq, PartialEq, Hash, Clone, Debug, Copy)]
+enum Key {
+    Account {
+        address: Address,
+        field_tag: AccountFieldTag,
+    },
+    AccountStorage {
+        tx_id: usize,
+        address: Address,
+        storage_key: Word,
+    },
+}
+
+impl Key {
+    fn address<F: Field>(&self) -> F {
+        match self {
+            Self::Account { address, .. } | Self::AccountStorage { address, .. } => {
+                address.to_scalar().unwrap()
+            }
+        }
+    }
+    fn field_tag<F: Field>(&self) -> F {
+        match self {
+            Self::Account { field_tag, .. } => F::from(*field_tag as u64),
+            Self::AccountStorage { .. } => F::zero(),
```

Comment:
Probably this function should be renamed to `proof_type` (following the update Miha mentioned).
And for AccountStorage, the result would be `F::from(MptProofType::StorageMod as u64)`